File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    graph = [[] for _ in range(n+1)]
    for a, b in edge:
        graph[a].append(b)
        graph[b].append(a)
        
    dp = [-1] * (n+1) # 1노드부터 n노드까지 거리
    dp[1] = 0
    
    q = deque([1])
    
    while q:
        now = q.popleft()
        for nxt in graph[now]:
            if dp[nxt] == -1:
                dp[nxt] = dp[now]+1
                q.append(nxt)
                
    max_dp = max(dp)
    return dp.count(max_dp)

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0;
        int max = -1;
        
        // 노드 연결
        List<Integer>[] graph = new ArrayList[n+1];
        for (int i = 0; i <= n; i++)
            graph[i] = new ArrayList<>();
        for (int[] e : edge) {
            graph[e[0]].add(e[1]);
            graph[e[1]].add(e[0]);
        }
        
        // 가장 먼 노드 찾기
        Queue<Integer> q = new ArrayDeque<>();
        int[] dist = new int[n+1]; // 1번 노드부터 i번 노드까지의 거리
        
        q.offer(1);
        dist[1] = 1;
        while(!q.isEmpty()) {
            int curr = q.poll();
            for (int neighbor : graph[curr]) {
                if (dist[neighbor] == 0) {
                    q.offer(neighbor);
                    dist[neighbor] = dist[curr] + 1;
                    if (dist[neighbor] > max) {
                        max = dist[neighbor];
                        answer = 1;
                    }
                    else if (dist[neighbor] == max)
                        answer++;
                }
            }
        }
            
        return answer;
    }
}